### ISSUE TITLE
C# fixed arrays not classified properly for the SysV AMD64 CC

### ISSUE THREAD

**Issue by pgavlin** [dotnet/coreclr#1974]
C# fixed arrays are represented in metadata as nested types of a fixed size with a single field of the fixed array element type. For example, this definition:

```
unsafe struct S
{
	public fixed float buf2[4];
}
```

results in this IL:

```
  .class private sequential ansi sealed beforefieldinit S
  	extends [System.Runtime]System.ValueType
  {
    .field  public  valuetype S/'<buf2>e__FixedBuffer' buf2
    .custom instance void class [mscorlib]System.Runtime.CompilerServices.FixedBufferAttribute::'.ctor'(class [mscorlib]System.Type, int32) =  (
		01 00 61 53 79 73 74 65 6D 2E 53 69 6E 67 6C 65   // ..aSystem.Single
		2C 20 53 79 73 74 65 6D 2E 52 75 6E 74 69 6D 65   // , System.Runtime
		2C 20 56 65 72 73 69 6F 6E 3D 34 2E 30 2E 32 30   // , Version=4.0.20
		2E 30 2C 20 43 75 6C 74 75 72 65 3D 6E 65 75 74   // .0, Culture=neut
		72 61 6C 2C 20 50 75 62 6C 69 63 4B 65 79 54 6F   // ral, PublicKeyTo
		6B 65 6E 3D 62 30 33 66 35 66 37 66 31 31 64 35   // ken=b03f5f7f11d5
		30 61 33 61 04 00 00 00 00 00                   ) // 0a3a......


  .class nested public sequential ansi sealed beforefieldinit '<buf2>e__FixedBuffer'
  	extends [System.Runtime]System.ValueType
  {
    .custom instance void class [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .custom instance void class [mscorlib]System.Runtime.CompilerServices.UnsafeValueTypeAttribute::'.ctor'() =  (01 00 00 00 ) // ....

    .pack 0
    .size 16
    .field  public  float32 FixedElementField

  } // end of class <buf2>e__FixedBuffer

  } // end of class S
```

The CoreCLR type system in turn represents this faithfully as a struct of size 16 with a single `float` field. Unfortunately, this causes the SysV classifier to drop the last 8 bytes of the struct: the classifier currently only classifies bytes that contain fields. Per ECMA-335 II.10.7, all bytes of the struct up to its explicit minimum size are significant (i.e. they are not treated like padding bytes in C/C++), so these leftover bytes must be classified. Furthermore, in order for C# fixed arrays to be compatible with C inline arrays (indeed, the intended purpose of C# fixed arrays is to represent C inline arrays), these bytes must be classified using the element type of the fixed array. After some discussion with @LLITCHEV, we propose that for any struct whose fields do not cover the entire size of the struct, the classification of the leftover trailing bytes is SSE if the classification of the preceeding eightbyte is SSE and INTEGER otherwise. As this situation has no direct analog in C/C++ (which generally treats such bytes as insignificant and drops them for the purposes of argument passing), this is a CLR-specific projection onto the SysV ABI that needs to be explicitly documented.

**Comment by pgavlin** [comment]
@LLITCHEV, @AndyAyersMS, @jkotas, @CarolEidt PTAL.

@jkotas, @CarolEidt: One specific question I have here regards ECMA-335 II.10.7: are the leftover bytes intended to be of any particular type (i.e. should they in fact be treated as if each byte was a field of type `byte`, which would change the classification)? If so, we'll need to be able to properly distinguish C# fixed buffers from other similar cases, as the leftover bytes have different semantics for C# fixed buffers. Fortunately, the FixedBufferAttribute gives us the opportunity to distinguish these fields, but the VM would need to be taught to parse and report this attribute (unless I've missed something and it already does).

**Comment by pgavlin** [comment]
One potential issue with the proposed approach is that it does not address holes at the beginning of a struct, e.g.:

```
[StructLayout(LayoutKind.Explicit, Size = 16)]
struct S
{
    [FieldOffset(8)]
    public int i;
}
```

Perhaps such bytes should be classified as INTEGER as a rule (in which case it would be kind of nice to be consistent and classify all leftover bytes in eightbytes that do not contain fields as INTEGER, contrary to the original proposal).

**Comment by LLITCHEV** [comment]
Per decision sometime ago explicit layout structs are always passed on stack.

**Comment by jkotas** [comment]
> explicit layout structs are always passed on stack

Right. It is the same what we do for HFAs on ARM/ARM64.

**Comment by pgavlin** [comment]
> explicit layout structs are always passed on stack

I don't think that this is a very user-friendly solution: how are developers intended to represent C unions in managed code? Just using the widest type for the field?

**Comment by pgavlin** [comment]
(note that in the case of managed code calling other managed code the aforementioned stack-only solution is fine. AFAIK, the example above is only representable in managed code (I know of no way to achieve the same effect in C/C++), so it's not as compelling in the context of P/Invokes)

**Comment by pgavlin** [comment]
Fixed with #1981.
